File: 240.search-a-2-d-matrix-ii.js
/*
 * @lc app=leetcode id=240 lang=javascript
 *
 * [240] Search a 2D Matrix II
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  let i = 0;
  let j = matrix[0].length - 1;
  while (i < matrix.length && j >= 0) {
    if (matrix[i][j] === target) {
      return true;
    }
    matrix[i][j] > target ? --j : ++i;
  }
  return false;
};
// @lc code=end

File: 240.search-a-2-d-matrix-ii.py
#
# @lc app=leetcode id=240 lang=python3
#
# [240] Search a 2D Matrix II
#

# Time:  O(m + n)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        i, j = 0, len(matrix[0]) - 1
        while i < len(matrix) and j >= 0:
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] > target:
                j -= 1
            else:
                i += 1
        return False


# @lc code=end
